netebpfext/net_ebpf_ext_hook_provider.c

```diff
@@ -0,0 +1,336 @@
+        KeSetEvent(&rundown->dispatch.rundown_wait, 0, FALSE);
+}
+
+/**
+ * @brief Initialize the hook client rundown state.
+ *
+ * @param[in, out] hook_client Pointer to the attached hook NPI client.
+ * @param execution_type Type of rundown support required.
+ *
+ * @retval STATUS_SUCCESS Operation succeeded.
+ * @retval STATUS_INSUFFICIENT_RESOURCES IO work item could not be allocated.
+ */
+static NTSTATUS
+_ebpf_ext_attach_init_rundown(net_ebpf_extension_hook_client_t* hook_client)
+{
+    NTSTATUS status = STATUS_SUCCESS;
+    net_ebpf_ext_hook_client_rundown_t* rundown = &hook_client->rundown;
+    net_ebpf_extension_hook_execution_t execution_type = hook_client->execution_type;
+
+    //
+    // Allocate work item for client detach processing.
+    //
+    hook_client->detach_work_item = IoAllocateWorkItem(_net_ebpf_ext_driver_device_object);
+    if (hook_client->detach_work_item == NULL) {
+        status = STATUS_INSUFFICIENT_RESOURCES;
+        goto Exit;
+    }
+
+    if (execution_type == EXECUTION_PASSIVE) {
+        ExInitializePushLock(&rundown->passive.lock);
+    } else {
+        KeInitializeEvent(&(rundown->dispatch.rundown_wait), SynchronizationEvent, FALSE);
+        KeInitializeDpc(&(rundown->dispatch.rundown_dpc), _ebpf_ext_attach_rundown, rundown);
+    }
+    rundown->rundown_occurred = FALSE;
+
+Exit:
+    return status;
+}
+
+/**
+ * @brief Block execution of the thread until all invocations are completed.
+ *
+ * @param[in, out] rundown Rundown object to wait for.
+ * @param[in]  execution_type Execution type for the hook (passive or dispatch).
+ *
+ */
+static void
+_ebpf_ext_attach_wait_for_rundown(
+    _Inout_ net_ebpf_ext_hook_client_rundown_t* rundown, net_ebpf_extension_hook_execution_t execution_type)
+{
+    rundown->rundown_occurred = TRUE;
+    if (execution_type == EXECUTION_PASSIVE) {
+        ExAcquirePushLockExclusive(&rundown->passive.lock);
+        ExReleasePushLockExclusive(&rundown->passive.lock);
+    } else {
+        // Queue a DPC to each CPU and wait for it to run.
+        // After it has run on each CPU we can be sure that no
+        // DPC is busy processing a hook.
+        uint32_t maximum_processor = KeQueryMaximumProcessorCount();
+        uint32_t processor;
+        for (processor = 0; processor < maximum_processor; processor++) {
+            KeSetTargetProcessorDpc(&rundown->dispatch.rundown_dpc, (uint8_t)processor);
+            if (KeInsertQueueDpc(&rundown->dispatch.rundown_dpc, NULL, NULL)) {
+                KeWaitForSingleObject(&rundown->dispatch.rundown_wait, Executive, KernelMode, FALSE, NULL);
+            }
+        }
+    }
+}
+
+IO_WORKITEM_ROUTINE _net_ebpf_extension_detach_client_completion;
+#pragma alloc_text(PAGE, _net_ebpf_extension_detach_client_completion)
+
+/**
+ * @brief IO work item routine callback that waits on client rundown to complete.
+ *
+ * @param[in] device_object IO Device object.
+ * @param[in] context Pointer to work item context.
```

Comment:
```suggestion
 * @param[in] context Pointer to work item context, if any.
```

Reply:
The SAL annotation of _in_opt_ comes from `IO_WORKITEM_ROUTINE` and cannot be changed. However in this case we assert that context is not null.